src/util/base.cc

```diff
@@ -0,0 +1,231 @@
+            reinterpret_cast<char*>(in_buffer),
+            input.gcount(),
+            out_buffer,
+            Size * 2);
+
+      output.write(reinterpret_cast<char*>(out_buffer), ret);
+      if (output.bad())
+        {
+          std::cerr << "Error : Output to stream failed !" << std::endl;
+          return false;
+        }
+      output.flush();
+    } while (input.good());
+  return true;
+}
+
+bool BaseCommand::Impl(
+    const std::string& cmd_name,
+    int argc,
+    const char* argv[])
+{
+  bpo::variables_map vm;
+  bpo::positional_options_description pos;
+  pos.add("type", 1).add("infile", 1).add("outfile", 1);
+  try
+    {
+      bpo::parsed_options parsed = bpo::command_line_parser(argc, argv)
+                                       .options(m_desc)
+                                       .positional(pos)
+                                       .run();
+      bpo::store(parsed, vm);
+      bpo::notify(vm);
+    }
+  catch (const bpo::error& e)
+    {
+      std::cerr << GetName() << " ERROR: " << e.what() << std::endl;
+      return false;
+    }
+
+  if (vm.count("help") || (argc < 2) || (argc > 4))
+    {
+      if (argc < 2)
+        std::cerr << "Not enough arguments !" << std::endl;
+      else if (argc > 4)
+        std::cerr << "Too many arguments !" << std::endl;
+      PrintUsage(cmd_name);
+      return false;
+    }
+
+  // encode or decode or error
+  bool encode = true;
+  if (m_opt_type == "decode")
+    {
+      encode = false;
+    }
+  else if (m_opt_type != "encode")
+    {
+      std::cerr << "Invalid option : \"" << m_opt_type << "\"" << std::endl;
+      std::cerr << "Should be \"encode\" or \"decode\"" << std::endl;
+      PrintUsage(cmd_name);
+      return false;
+    }
+
+  // Streams management
+  std::shared_ptr<std::istream> istream;
+  if (m_opt_infile.empty() || m_opt_infile == "-")
+    {  // read from stdin
+      istream.reset(&std::cin, [](...) {});  // noop : don't delete
+    }
+  else
+    {  // read from file
+      std::ifstream* ifs = new std::ifstream();
+      ifs->open(m_opt_infile.c_str(), std::ios::in | std::ios::binary);
+      if (ifs->fail())
+        {
+          std::cerr << "Failed to open input file \"" << m_opt_infile.c_str()
+                    << "\"" << std::endl;
+          delete ifs;
+          return false;
+        }
+      istream.reset(ifs);
+    }
+
+  std::shared_ptr<std::ostream> ostream;
+  if (m_opt_outfile.empty() || m_opt_outfile == "-")
+    {  // write to stdout
+      ostream.reset(&std::cout, [](...) {});  // noop : don't delete
+    }
+  else
+    {  // write to file
+      std::ofstream* ofs = new std::ofstream();
+      ofs->open(m_opt_outfile.c_str(), std::ios::out | std::ios::binary);
+      if (ofs->fail())
+        {
+          std::cerr << "Failed to open output file \"" << m_opt_outfile.c_str()
+                    << "\"" << std::endl;
+          delete ofs;
+          return false;
+        }
+      ostream.reset(ofs);
+    }
```

Comment:
>&anonimal       moroccanmalinois: IMHO this is great. I think you'll refactor/reuse the file/stream handling code as you progress so it's not endemic only to BaseCommand? Could we consider implementing it/or related in kovri::core src/core/util/filesystem?

@MoroccanMalinois said he will move stream handling to `src/core/util/filesystem` 👍.

Note: once that's done, if re-implemented, I think we can use it throughout the codebase 😄